### ISSUE TITLE
Unable to reproduce FLAIR paper results with provided configuration

### ISSUE THREAD

**Issue by showstarpro** [ExplainableML/flair#6]
Hi FLAIR team,

I'm trying to reproduce the results using the configuration from the paper, but getting significantly poorer performance (near-zero retrieval metrics). Here are my observations:

torchrun --nnode 1 --nproc_per_node 8 -m main \
    --model ViT-B-16-FLAIR \
    --train-data 'dataset/cc3m-3long-3short-1raw/0-1-0/cc3m_3long_3short_1raw_captions/00{000..287}.tar' \
    --train-num-samples  2823019 \
    --train-dataset-type webdataset  \
    --use-flair-loss \
    --add-mps-loss \
    --num-sampled-captions 8 \
    --log-every-n-steps 200 \
    --caption-sampling-mode diverse_sampling \
    --epochs 32 \
    --batch-size 128 \
    --lr 5e-4 \
    --warmup 2000 \
    --precision amp \
    --workers 25 \
    --delete-previous-checkpoint \
    --beta1 0.9 \
    --beta2 0.98 \
    --wd 0.5 \
    --eps 1e-8

Evaluation metrics for epoch 32:  retrieval_coco_text_to_image_R@1: 0.0002 (vs. paper's ~40+ R@1)
Full metrics: 
[ours]: 2025-04-03,02:05:25 | INFO | Eval Epoch: 0 retrieval_coco_text_to_image_R@1: 0.0002     retrieval_coco_text_to_image_R@5: 0.0010        retrieval_coco_text_to_image_R@10: 0.0020     retrieval_coco_text_to_image_mean_rank: 2502.1619       retrieval_coco_text_to_image_median_rank: 2503.0000     retrieval_coco_image_to_text_R@1: 0.0000      retrieval_coco_image_to_text_R@5: 0.0010        retrieval_coco_image_to_text_R@10: 0.0018       retrieval_coco_image_to_text_mean_rank: 4344.3032       retrieval_coco_image_to_text_median_rank: 3365.0000   epoch: 0.0000   retrieval_coco_num_text_samples: 25014.0000     retrieval_coco_num_image_samples: 5000.0000
[your checkpoint]: 2025-04-03,02:02:33 | INFO | Eval Epoch: 0 retrieval_coco_text_to_image_R@1: 0.3765     retrieval_coco_text_to_image_R@5: 0.6444        retrieval_coco_text_to_image_R@10: 0.7488     retrieval_coco_text_to_image_mean_rank: 19.0804 retrieval_coco_text_to_image_median_rank: 3.0000        retrieval_coco_image_to_text_R@1: 0.5162     retrieval_coco_image_to_text_R@5: 0.7716 retrieval_coco_image_to_text_R@10: 0.8504       retrieval_coco_image_to_text_mean_rank: 8.1024  retrieval_coco_image_to_text_median_rank: 1.0000      epoch: 0.0000   retrieval_coco_num_text_samples: 25014.0000     retrieval_coco_num_image_samples: 5000.0000

Are there any critical configuration differences between the paper and this command?
Would appreciate any guidance on potential configuration issues or overlooked implementation details. Thank you!

[out.log](https://github.com/user-attachments/files/19578724/out.log)

**Comment by sean-xr** [comment]
Hi, 

Thanks for trying the training out! It's indeed wierd because from your logs, it seems that the loss is not decreasing at all. In my case, the training loss should decrease to 1.x. Let me check it carefully and reproduce the training again. 

[output_cc3m.log](https://github.com/user-attachments/files/19583691/output_cc3m.log)

**Comment by showstarpro** [comment]
Thank you sincerely for your prompt response and for sharing your valuable insights! I truly appreciate your support.

Regarding the training process, I observed that the loss in my logs does not decrease at all, whereas you mentioned it should reduce to around 1.x. To better align my setup with yours, could you kindly clarify the following:

Training Script Differences :
Are there any specific configurations, hyperparameters (e.g., learning rate, optimizer settings), or preprocessing steps in your script that might differ from the ones I provided? 
Dataset Details :
I obtained the cc3m_3long_3short_1raw_captions dataset by downloading CSV files from qidouxiong619/dreamlip_long_captions and converting them to WebDataset format via img2dataset, resulting in 2,823,019 image-text pairs.
Could you confirm if this matches your data pipeline? 
I am keen to resolve the training discrepancy and ensure our setups are compatible. Any further details or suggestions would be immensely helpful!

Thank you again for your time and guidance.

**Comment by sean-xr** [comment]
Ah I see the difference. In FLAIR, the captions are splitted the same way as DreamLIP does:

<pre> \```python def split_caption(text): texts = re.split(r'\n|</s>|[.]', text) subcap = [] for text_prompt in texts: text_prompt = text_prompt.strip() if len(text_prompt) != 0: subcap.append(text_prompt) del texts return subcap \``` </pre>

After scraping the dataset using `img2dataset`, I splitted all the captions using this function. If you don't want to modify the shards now, you could also apply this function in the `sample_dict()' function in `src/data.py` to split all the captions:


<pre> \```python captions_list = (split_caption(text['raw_caption']) + split_caption(text['shortIB_captions']) + ... + split_caption(text['longLLA_captions'])) \``` </pre>

Sorry for not having pointed this out, I will update the README. Could you see if this addresses your issue first?

Thanks for your time as well!

**Comment by showstarpro** [comment]
I will try. Thanks for your help!!!

**Comment by sean-xr** [comment]
Hello, thank you so much for trying the training out again, I'm happy that the pre-processing addresses your issue, I will also update the README soon so people don't get confused.

However, `2823019` is the number of valida samples that I get in my CC3M dataset, are you sure you also get this number after your downlaod? I'm pretty sure you will get less due to the rotten links : ). In `webdataset`, a valid image-text pair is recognized as the image and caption with the same name, e.g. `000012.jpg` and `000012.json`. So could you iterate through all your shards and count the valid number of pairs? if that's lower than `2823019`, then it should explain the performance gap.

Thank you so much for your efforts. You have indeed made this repo better!

**Comment by showstarpro** [comment]
I have just reviewed the JSON logs from the img2dataset download process and found that only 1,711,097 samples were successfully downloaded. This represents a significant gap from the target of 2,823,019 samples. Thank you for your reminder, and I would like to commend this high-quality work again.

**Comment by sean-xr** [comment]
Thanks for the nice words. Feel free to reopen this issue again if you have any questions.
